### PR TITLE
core: cpu: sys: simplify reboot()

### PR DESCRIPTION
**Pull request by kaspar030**
This PR changes the signature of reboot to ```void reboot(void)```

- reboot() returning a value is a bug
- reboot()'s mode parameter was unused

(also changes _exit printf to "LOG_INFO")

### CODE REVIEW

**Comment by kaspar030** [comment]
kaspar@ifive:~/src/riot/examples/default$ OLDBIN=bin NEWBIN=newbin make info-buildsizes-diff
text	data	bss	dec	BOARD/BINDIRBASE

-88	0	0	-88	airfy-beacon
20400	164	10316	30880	bin
20312	164	10316	30792	newbin

-76	0	0	-76	arduino-due
9936	164	2660	12760	bin
9860	164	2660	12684	newbin

-50	-28	0	-78	arduino-mega2560
7264	676	705	8645	bin
7214	648	705	8567	newbin

-40	0	0	-40	avsextrem
48236	2340	95961	146537	bin
48196	2340	95961	146497	newbin

-76	0	0	-76	cc2538dk
9664	160	2624	12448	bin
9588	160	2624	12372	newbin

ERR	ERR	ERR	ERR	chronos
ERR	ERR	ERR	ERR	bin
ERR	ERR	ERR	ERR	newbin

-80	0	0	-80	ek-lm4f120xl
9988	160	2624	12772	bin
9908	160	2624	12692	newbin

-80	0	0	-80	f4vi1
11300	164	2620	14084	bin
11220	164	2620	14004	newbin

-80	0	0	-80	fox
26120	164	10484	36768	bin
26040	164	10484	36688	newbin

-80	0	0	-80	frdm-k64f
14444	1256	4800	20500	bin
14364	1256	4800	20420	newbin

-80	0	0	-80	iotlab-m3
33576	164	11612	45352	bin
33496	164	11612	45272	newbin

-80	0	0	-80	limifrog-v1
11056	160	2632	13848	bin
10976	160	2632	13768	newbin

-76	0	0	-76	mbed_lpc1768
9780	160	2632	12572	bin
9704	160	2632	12496	newbin

ERR	ERR	ERR	ERR	msb-430
ERR	ERR	ERR	ERR	bin
ERR	ERR	ERR	ERR	newbin

ERR	ERR	ERR	ERR	msb-430h
ERR	ERR	ERR	ERR	bin
ERR	ERR	ERR	ERR	newbin

-48	0	0	-48	msba2
61396	2348	93965	157709	bin
61348	2348	93965	157661	newbin

-80	0	0	-80	msbiot
11580	164	2644	14388	bin
11500	164	2644	14308	newbin

-84	0	0	-84	mulle
35584	1280	13360	50224	bin
35500	1280	13360	50140	newbin

-120	0	0	-120	native
69291	612	82900	152803	bin
69171	612	82900	152683	newbin

-88	0	0	-88	nrf51dongle
20424	164	10316	30904	bin
20336	164	10316	30816	newbin

-88	0	0	-88	nrf6310
20400	164	10316	30880	bin
20312	164	10316	30792	newbin

-88	0	0	-88	nucleo-f091
16760	164	2644	19568	bin
16672	164	2644	19480	newbin

-80	0	0	-80	nucleo-f303
11368	160	2648	14176	bin
11288	160	2648	14096	newbin

-80	0	0	-80	nucleo-f334
11112	160	2624	13896	bin
11032	160	2624	13816	newbin

-80	0	0	-80	nucleo-l1
10940	160	2624	13724	bin
10860	160	2624	13644	newbin

-76	0	0	-76	openmote
9572	160	2624	12356	bin
9496	160	2624	12280	newbin

-80	0	0	-80	pba-d-01-kw2x
32112	1256	13192	46560	bin
32032	1256	13192	46480	newbin

-88	0	0	-88	pca10000
20424	164	10316	30904	bin
20336	164	10316	30816	newbin

-88	0	0	-88	pca10005
20404	164	10316	30884	bin
20316	164	10316	30796	newbin

-40	0	0	-40	pttu
48444	2340	95961	146745	bin
48404	2340	95961	146705	newbin

-99	0	28	-71	qemu-i386
124394	5528	68164	198086	bin
124295	5528	68192	198015	newbin

-76	0	0	-76	remote
10228	160	2624	13012	bin
10152	160	2624	12936	newbin

-88	0	0	-88	saml21-xpro
20612	164	10444	31220	bin
20524	164	10444	31132	newbin

-88	0	0	-88	samr21-xpro
30612	168	11592	42372	bin
30524	168	11592	42284	newbin

-80	0	0	-80	spark-core
22008	164	10444	32616	bin
21928	164	10444	32536	newbin

-88	0	0	-88	stm32f0discovery
13096	160	2632	15888	bin
13008	160	2632	15800	newbin

-80	0	0	-80	stm32f3discovery
11372	160	2648	14180	bin
11292	160	2648	14100	newbin

-80	0	0	-80	stm32f4discovery
11464	164	2636	14264	bin
11384	164	2636	14184	newbin

ERR	ERR	ERR	ERR	telosb
ERR	ERR	ERR	ERR	bin
ERR	ERR	ERR	ERR	newbin

-76	0	0	-76	udoo
9936	164	2660	12760	bin
9860	164	2660	12684	newbin

ERR	ERR	ERR	ERR	wsn430-v1_3b
ERR	ERR	ERR	ERR	bin
ERR	ERR	ERR	ERR	newbin

ERR	ERR	ERR	ERR	wsn430-v1_4
ERR	ERR	ERR	ERR	bin
ERR	ERR	ERR	ERR	newbin

-88	0	0	-88	yunjia-nrf51822
20388	164	10316	30868	bin
20300	164	10316	30780	newbin

ERR	ERR	ERR	ERR	z1
ERR	ERR	ERR	ERR	bin
ERR	ERR	ERR	ERR	newbin

**Comment by OlegHahm** [comment]
I think this the third PR with this change.

**Comment by OlegHahm** [comment]
See #822.

**Comment by kaspar030** [comment]
So we were willing to pay 80 bytes of code to have a BSD compatible reboot call?

We should reconsider.

**Comment by gebart** [comment]
I don't understand what was costing 80 bytes. Is it removing the printf message that gives the space savings?

**Comment by kaspar030** [comment]
The message is ~30bytes. The delay loop was quite a bit, another layer of function calls, register juggling, ...

**Comment by OlegHahm** [comment]
Actually, I was lately thinking about using the parameter to introduce some kind of soft-reboot (restart RIOT without resetting the device).

**Comment by kaspar030** [comment]
That function should go into "soft_reboot()" or something similar (if we find a use-case).

**Comment by kaspar030** [comment]
- refactored reboot() so OS X builds again

**Comment by OlegHahm** [comment]
One use-case is bringing them system in a defined state without losing all information between the reboot.

**Comment by kaspar030** [comment]
@OlegHahm OK, interesting, but still that should get it's own function, right?

**Comment by OlegHahm** [comment]
Which would probably waste more ROM.

**Comment by kaspar030** [comment]
@OlegHahm bullshit.

**Comment by kaspar030** [comment]
*bump*

I still think this change is overdue.

- there's no use case for *BSD compatibility other than avoiding native header clashes (#822), which this PR avoids

- implementing @OlegHahm's proposal to use the parameter to do a soft reset can more efficiently implemented as

```
void flexible_reboot(parameter) {
  if (parameter==SOFT) {
    reboot_soft();
  else {
    reboot();
}
```

thus making the extra overhead optional.

(This PR is basically the same as #3246, it just skips reboot_arch indirection.)

File: sys/newlib/syscalls.c [hunk 1]
```diff
@@ -71,8 +72,8 @@ void _fini(void)
  */
 void _exit(int n)
 {
-    printf("#! exit %i: resetting\n", n);
-    reboot(n);
+    LOG_INFO("#! exit %i: resetting\n", n);
```
Comment (OlegHahm) [inline sys/newlib/syscalls.c]: Isn't this message likely to get cut if logging to UART because of the removed delay loop?

Comment (kaspar030) [inline sys/newlib/syscalls.c]: Probably. That probably also was the case before on some platforms. We could xtimer_wait() this out, but I'd then rather do it only with DEVELHELP. What do you think?

Comment (OlegHahm) [inline sys/newlib/syscalls.c]: Hm, what happens if you're logging over UDP or on a memory card, for instance? I think if we use logging here, we should make sure that the reboot event actually gets logged.

Comment (kaspar030) [inline sys/newlib/syscalls.c]: IMHO that should be dealt with at the loggin layer (LOG_FLUSH() or similar).

**Comment by kaspar030** [comment]
- rebased

**Comment by kaspar030** [comment]
- fixed weio

**Comment by gebart** [comment]
```
/home/travis/build/RIOT-OS/RIOT/core/panic.c: In function 'core_panic':
/home/travis/build/RIOT-OS/RIOT/core/panic.c:75:5: error: implicit declaration of function 'reboot' [-Werror=implicit-function-declaration]
     reboot();
     ^
```

File: core/include/reboot.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
```
Comment (haukepetersen) [inline core/include/reboot.h]: We have 2016 :-)

File: cpu/atmega2560/reboot_arch.c [hunk 3]
```diff
@@ -14,27 +15,21 @@
...
+void reboot(void)
 {
-    printf("Going into reboot, mode %i\n", mode);
-
     /*
      * Since the AVR doesn't support a real software reset, we set the Watchdog
      * Timer on a 250ms timeout. Consider this a kludge.
      */
     wdt_enable(WDTO_250MS);
-
```
Comment (haukepetersen) [inline cpu/atmega2560/reboot_arch.c]: shouldn't there be an `while(1);` loop after this, so the execution stays here until the watchdog times out?

**Comment by haukepetersen** [comment]
big +1!

File: cpu/atmega2560/reboot_arch.c (hunk 3 above)
Comment (kaspar030) [inline cpu/atmega2560/reboot_arch.c]: yep... didn't touch the semantics. Will fix it anyways.

**Comment by kaspar030** [comment]
- addressed comments

File: cpu/atmega2560/reboot_arch.c (hunk 3 above)
Comment (haukepetersen) [inline cpu/atmega2560/reboot_arch.c]: nice

**Comment by haukepetersen** [comment]
Just thinking out loud: You dropped the `_arch` stuff, which indeed makes the code look nicer. But the major idea behind the `xxx_arch` was to clearly mark the dependencies between the kernel and hardware specific code. While don't really insist on this, this might be a good point in time to re-think that approach?! But let's take that discussion somewhere else -> #4669

**Comment by kaspar030** [comment]
Travis is happy (apart from squashing). @gebart @haukepetersen would you ACK this?

File: cpu/native/native_cpu.c [hunk 4]
```diff
@@ -1,19 +1,21 @@
 /**
- * Native CPU kernel_intern.h and sched.h implementation
- *
- * in-process preemptive context switching utilizes POSIX ucontexts.
- * (ucontext provides for architecture independent stack handling)
- *
- * Copyright (C) 2013 Ludwig Knüpfer <ludwig.knuepfer@fu-berlin.de>
+ * Copyright (C) 2016 Kaspar Schleiser <kaspar@schleiser.de>
```
Comment (gebart) [inline cpu/native/native_cpu.c]: this should not be inside a normal comment (`/* */`)

**Comment by gebart** [comment]
Tested on Mulle, works. One minor comment unrelated to the actual change, fix it if you feel like it.

OK to squash.

1st ACK.

**Comment by kaspar030** [comment]
- fixed @gebart, squashed

**Comment by haukepetersen** [comment]
ACK and go.

**Comment by kaspar030** [comment]
Thanks for reviewing!

**Pull request merged**